spring-kafka/src/main/java/org/springframework/kafka/config/KafkaStreamsInfrastructureCustomizer.java

```diff
@@ -0,0 +1,47 @@
+/*
+ * Copyright 2020 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.kafka.config;
+
+import org.apache.kafka.streams.StreamsBuilder;
+import org.apache.kafka.streams.Topology;
+
+/**
+ * A customizer for infrastructure components such as the builder and topology.
+ *
+ * @author Gary Russell
+ * @since 2.4.1
+ *
+ */
+public interface KafkaStreamsInfrastructureCustomizer {
+
+	/**
+	 * Configure the builder.
+	 * @param builder the builder.
+	 */
+	default void configureBuilder(StreamsBuilder builder) {
+		// no-op
+	}
+
+	/**
+	 * Configure the topology.
+	 * @param topology the topology
+	 */
+	default void configureTopology(Topology topology) {
```

Comment:
I wonder if with this new API we ever need a separate `KafkaStreamsCustomizer` strategy.
Couldn't that be covered in this one as a separate `configureKafkaStreams(KafkaStreams)` callback?

Might be a story for other version, although we could deprecate now and suggest to move to a new API...

Reply:
Yeah - I thought about deprecating that, but I couldn't come up with a good name for a new all-encompassing interface and I don't want to break the `@FunctionalInterface` there in a point release and I don't want to force people who only want to customize the infrastructure to have to implement a `configureStream` method.

We can think about it for a future release.

Reply:
Yeah... I know what you mean.
How about otherwise then: pursue fine-graining?
Since we already have a `KafkaStreamsCustomizer` and `StreamsBuilderFactoryBeanCustomizer` it won't hurt to have to more `@FunctionalInterface`s instead of this not consistent strategy?

Reply:
Too many customizers, I think (we already have too many; I'd prefer just two, but can't do this in a point release). We can reconsider in 2.5 (or 3.0), where I would like to do a complete overhaul of the streams support.

Reply:
Sure! Let's postpone such a discussion!